plugins/out_doris/doris.c

```diff
@@ -0,0 +1,545 @@
+            
+            out_ret = http_put(ctx, redirect_host, atoi(redirect_port), 
+                               body, body_len, tag, tag_len, label, label_len, "be");
+        }
+        else if (c->resp.status == 200 && c->resp.payload_size > 0) {
+            ret = flb_pack_json(c->resp.payload, c->resp.payload_size,
+                                &out_buf, &out_size, &root_type, NULL);
+
+            if (ret == -1) {
+                out_ret = FLB_RETRY;
+                goto parse_done;
+            }
+            
+            msgpack_unpacked_init(&result);
+            ret = msgpack_unpack_next(&result, out_buf, out_size, &off);
+            if (ret != MSGPACK_UNPACK_SUCCESS) {
+                out_ret = FLB_RETRY;
+                goto free_buf;
+            }
+
+            root = result.data;
+            if (root.type != MSGPACK_OBJECT_MAP) {
+                out_ret = FLB_RETRY;
+            }
+
+            for (i = 0; i < root.via.map.size; i++) {
+                msg_key = root.via.map.ptr[i].key;
+                if (msg_key.type != MSGPACK_OBJECT_STR) {
+                    out_ret = FLB_RETRY;
+                    break;
+                }
+
+                if (msg_key.via.str.size == 6 && strncasecmp(msg_key.via.str.ptr, "Status", 6) == 0) {
+                    msg_val = root.via.map.ptr[i].val;
+                    if (msg_val.type != MSGPACK_OBJECT_STR) {
+                        out_ret = FLB_RETRY;
+                        break;
+                    }
+
+                    if (msg_val.via.str.size == 7 && strncasecmp(msg_val.via.str.ptr, "Success", 7) == 0) {
+                        out_ret = FLB_OK;
+                        break;
+                    }
+                    
+                    if (msg_val.via.str.size == 15 && strncasecmp(msg_val.via.str.ptr, "Publish Timeout", 15) == 0) {
+                        out_ret = FLB_OK;
+                        break;
+                    }
+                    
+                    out_ret = FLB_RETRY;
+                    break;
+                }
+            }
+free_buf:
+            flb_free(out_buf);
+            msgpack_unpacked_destroy(&result);
+parse_done:;
+        }
+        else {
+            out_ret = FLB_RETRY;
+        }
+    }
+    else {
+        flb_plg_error(ctx->ins, "could not flush records to %s:%i (http_do=%i)",
+                      ctx->host, ctx->port, ret);
+        out_ret = FLB_RETRY;
+    }
+
+    /* cleanup */
+    
+    /*
+     * If the payload buffer is different than incoming records in body, means
+     * we generated a different payload and must be freed.
+     */
+    if (payload_buf != body) {
+        flb_free(payload_buf);
+    }
+
+    /* Destroy HTTP client context */
+    flb_http_client_destroy(c);
+
+    /* Release the TCP connection */
+    flb_upstream_conn_release(u_conn);
+
+    return out_ret;
+}
+
+static int compose_payload(struct flb_out_doris *ctx,
```

Comment:
Why do you use `void *` for out_body and cast it everwhere else instead of using `unsigned char *` and casting it once?

Please try to make the code simple and easy to read in the future, that'll help others catch issues and not make them themselves.